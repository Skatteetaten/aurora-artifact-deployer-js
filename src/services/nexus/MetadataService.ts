import crypto from 'crypto';

import {
  PROJECT_METADATA_XML,
  LATEST_METADATA_XML,
  POM_XML,
  MavenSchema
} from './templates';

export interface ContentMetaData {
  path: string;
  name: string;
}

export interface ContentTypes {
  artifact: ContentMetaData;
  project: ContentMetaData;
  pom: ContentMetaData;
  latest?: ContentMetaData;
}

export type FileType = keyof ContentTypes;

export interface UploadFile {
  path: string;
  name: string;
  type: FileType;
  content: Buffer;
}

export class MetadataService {
  private schema: MavenSchema;

  public constructor(schema: MavenSchema) {
    this.schema = schema;
  }

  public generateMetadataFiles(
    artifactContent: Buffer,
    classifier?: string
  ): UploadFile[] {
    const meta = this.getUploadMetadata(classifier);

    const artifactFiles = this.generateUploadFiles(
      'artifact',
      meta.artifact,
      artifactContent
    );
    const pomFiles = this.generateUploadFiles(
      'pom',
      meta.pom,
      POM_XML(this.schema)
    );
    const projectFiles = this.generateUploadFiles(
      'project',
      meta.project,
      PROJECT_METADATA_XML(this.schema)
    );

    const latestFiles = meta.latest
      ? this.generateUploadFiles(
          'latest',
          meta.latest,
          LATEST_METADATA_XML(this.schema)
        )
      : [];

    return [...artifactFiles, ...projectFiles, ...pomFiles, ...latestFiles];
  }

  private getUploadMetadata(classifier?: string): ContentTypes {
    const { artifactId, groupId, version, packaging } = this.schema;

    const applyClassifier = (text: string): string => {
      return classifier ? `${text}-${classifier}` : text;
    };

    const groupIdAsPath = groupId.replace(/\./g, '/');
    const artifactName = artifactId + '-' + version;
    const artifactNameWithClassifier = applyClassifier(artifactName);

    const files: ContentTypes = {
      project: {
        path: `${groupIdAsPath}/${artifactId}`,
        name: `maven-metadata.xml`
      },
      pom: {
        path: `${groupIdAsPath}/${artifactId}/${version}`,
        name: `${artifactName}.pom`
      },
      artifact: {
        path: `${groupIdAsPath}/${artifactId}/${version}`,
        name: `${artifactNameWithClassifier}.${packaging}`
      }
    };

    if (version.search(/.*SNAPSHOT$/i) !== -1) {
      files.latest = {
        path: `${groupIdAsPath}/${artifactId}/${version}`,
        name: `maven-metadata.xml`
      };
    }

    return files;
  }

  private generateUploadFiles(
    type: FileType,
    { name, path }: ContentMetaData,
    content: Buffer
  ): UploadFile[] {
    return [
      { path, name, content, type },
      {
        path,
        type,
        name: `${name}.md5`,
        content: this.md5(content)
      },
      {
        path,
        type,
        name: `${name}.sha1`,
        content: this.sha1(content)
      }
    ];
  }

  private md5(buf: Buffer): Buffer {
    const hash = crypto.createHash('md5');
    const content = hash.update(buf).digest('hex');
    return Buffer.from(content);
  }

  private sha1(buf: Buffer): Buffer {
    const hash = crypto.createHash('sha1');
    const content = hash.update(buf).digest('hex');
    return Buffer.from(content);
  }
}
