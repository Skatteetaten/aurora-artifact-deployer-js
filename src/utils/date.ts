// Date format: yyyymmddHHMMss
export function getLastUpdatedTime(date = new Date()): string {
  const withZero = (n: number): string => `0${n}`.slice(-2);

  const rest = [
    date.getMonth() + 1,
    date.getDate(),
    date.getHours(),
    date.getMinutes(),
    date.getSeconds()
  ].map((n): string => withZero(n));

  return [date.getFullYear(), ...rest].join('');
}
