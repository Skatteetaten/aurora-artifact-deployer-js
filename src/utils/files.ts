import fs from 'fs';

function directoryExists(dir: string): boolean {
  try {
    return fs.statSync(dir).isDirectory();
  } catch (e) {
    const error = e as NodeJS.ErrnoException;
    // error is thrown by statSync when path does not exist
    if (error.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
}

export function mkdirp(dir: string): void {
  dir
    .split('/')
    .reduce((paths: string[], next: string): string[] => {
      return paths.length === 0
        ? [next]
        : [
            ...paths,
            paths
              .slice(paths.length - 1)
              .concat(next)
              .join('/'),
          ];
    }, [])
    .forEach((path): void => {
      if (!directoryExists(path)) {
        try {
          fs.mkdirSync(path);
        } catch (e) {
          console.log(e);
        }
      }
    });
}
