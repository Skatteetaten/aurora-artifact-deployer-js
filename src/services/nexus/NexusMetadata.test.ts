import { MetadataService } from './MetadataService';
import { getLastUpdatedTime } from '../../utils/date';

describe('MetadataService', () => {
  describe('generateMetadataFiles', () => {
    it('should create files', () => {
      const metadataService = new MetadataService({
        artifactId: 'deployer',
        groupId: 'no.skatteetaten.aurora',
        lastUpdated: getLastUpdatedTime(new Date('August 26, 2019 20:08:00')),
        packaging: 'tgz',
        version: '1.0.0'
      });

      const files = metadataService.generateMetadataFiles(
        Buffer.from('test'),
        'Webleveransepakke'
      );
      const filesWithStringContent = files.map(file => ({
        ...file,
        content: file.content.toString()
      }));

      expect(filesWithStringContent).toMatchSnapshot();
    });
  });
});
