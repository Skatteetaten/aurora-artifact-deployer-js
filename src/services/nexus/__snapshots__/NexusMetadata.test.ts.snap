// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataService generateMetadataFiles should create files 1`] = `
Array [
  Object {
    "content": "test",
    "name": "deployer-1.0.0-Webleveransepakke.tgz",
    "path": "no/skatteetaten/aurora/deployer/1.0.0",
    "type": "artifact",
  },
  Object {
    "content": "098f6bcd4621d373cade4e832627b4f6",
    "name": "deployer-1.0.0-Webleveransepakke.tgz.md5",
    "path": "no/skatteetaten/aurora/deployer/1.0.0",
    "type": "artifact",
  },
  Object {
    "content": "a94a8fe5ccb19ba61c4c0873d391e987982fbbd3",
    "name": "deployer-1.0.0-Webleveransepakke.tgz.sha1",
    "path": "no/skatteetaten/aurora/deployer/1.0.0",
    "type": "artifact",
  },
  Object {
    "content": "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<metadata>
  <groupId>no.skatteetaten.aurora</groupId>
  <artifactId>deployer</artifactId>
  <versioning>
    <latest>1.0.0</latest>
    <versions>
      <version>1.0.0</version>
    </versions>
    <lastUpdated>20190826200800</lastUpdated>
  </versioning>
</metadata>
",
    "name": "maven-metadata.xml",
    "path": "no/skatteetaten/aurora/deployer",
    "type": "project",
  },
  Object {
    "content": "309c0689fbb0da2223dea252f6f60e78",
    "name": "maven-metadata.xml.md5",
    "path": "no/skatteetaten/aurora/deployer",
    "type": "project",
  },
  Object {
    "content": "7575a4e8a8fbb75778ea4912d56b40246105f8bc",
    "name": "maven-metadata.xml.sha1",
    "path": "no/skatteetaten/aurora/deployer",
    "type": "project",
  },
  Object {
    "content": "<project xmlns=\\"http://maven.apache.org/POM/4.0.0\\"
  xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\"
  xsi:schemaLocation=\\"http://maven.apache.org/POM/4.0.0
                      http://maven.apache.org/xsd/maven-4.0.0.xsd\\">
  <modelVersion>4.0.0</modelVersion>
  <groupId>no.skatteetaten.aurora</groupId>
  <artifactId>deployer</artifactId>
  <version>1.0.0</version>
  <packaging>tgz</packaging>
</project>
",
    "name": "deployer-1.0.0.pom",
    "path": "no/skatteetaten/aurora/deployer/1.0.0",
    "type": "pom",
  },
  Object {
    "content": "2ae5c8c90c512a756bad34f6c603c9a5",
    "name": "deployer-1.0.0.pom.md5",
    "path": "no/skatteetaten/aurora/deployer/1.0.0",
    "type": "pom",
  },
  Object {
    "content": "0d6ace405906a7528d555fa04f415135c8861911",
    "name": "deployer-1.0.0.pom.sha1",
    "path": "no/skatteetaten/aurora/deployer/1.0.0",
    "type": "pom",
  },
]
`;
